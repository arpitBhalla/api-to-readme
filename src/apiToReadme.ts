import fs from "fs";

export type API = {
  name: string;
  description?: string;
  endpoint: string;
  method: "POST" | "GET" | "DELETE" | "PUT";
  request?: unknown;
  response: unknown;
};

export type Options = {
  readmePath?: string;
  startComment?: string;
  endComment?: string;
};

export const makeReadme = (
  apis: API[],
  options: Options = {}
): string | void => {
  const { readmePath, startComment = "api", endComment = "api-end" } = options;
  const md = apis
    .map((api) => {
      const request = api.request
        ? `\n**Request Body**\n
${"```json"}
${JSON.stringify(api.request, null, 2)}
${"```"}\n`
        : "";
      return `### **${api.name}**
${api.description ? `\n${api.description}\n` : ""}
> **\`${api.method}\`**  **/${api.endpoint}**
${request}
**Response**\n
${"```json"}
${JSON.stringify(api.response, null, 4)}
${"```"}
`;
    })
    .join("\n");
  if (!readmePath) return md;
  const oldReadme = fs.readFileSync(readmePath, "utf8");
  const commentStart = new RegExp(`<!--\\s*${startComment}\\s*-->`).exec(
    oldReadme
  );
  const commentEnd = new RegExp(`<!--\\s*${endComment}\\s*-->`).exec(oldReadme);
  if (commentStart && commentEnd && commentEnd.index > commentStart.index) {
    const newReadme = `${oldReadme.substring(
      0,
      commentStart.index + commentStart[0].length
    )}
${md}
${oldReadme.substring(commentEnd.index)}`;
    if (newReadme !== oldReadme) {
      fs.writeFileSync(readmePath, newReadme, "utf8");
    }
    return newReadme;
  }
};
