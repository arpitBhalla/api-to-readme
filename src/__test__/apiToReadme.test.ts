import { makeReadme, API } from "../apiToReadme";
import path from "path";

describe("Name of the group", () => {
  const apis: API[] = [
    {
      name: "Search",
      description: "Search for user",
      endpoint: "search?q={string}&page={number}&perPage={number}",
      method: "GET",
      response: {
        code: 200,
        posts: [
          {
            name: "string",
          },
          {
            name: "string",
          },
        ],
      },
    },
    {
      name: "New User",
      endpoint: "newUser",
      method: "POST",
      request: {
        name: "string",
        password: "string",
      },
      response: {
        code: 200,
      },
    },
  ];

  it("should ", () => {
    const readme = makeReadme(apis);
    expect(readme).toMatchSnapshot();
  });

  it("should ", () => {
    const readme = makeReadme(apis, {
      readmePath: path.join(__dirname, "readme.md"),
      startComment: "api-start",
      endComment: "api-end",
    });
    expect(readme).toMatchSnapshot();
  });
});
