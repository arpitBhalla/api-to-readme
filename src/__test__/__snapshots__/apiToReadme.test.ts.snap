// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Name of the group should  1`] = `
"### **Search**

Search for user

> **\`GET\`**  **/search?q={string}&page={number}&perPage={number}**

**Response**

\`\`\`json
{
    \\"code\\": 200,
    \\"posts\\": [
        {
            \\"name\\": \\"string\\"
        },
        {
            \\"name\\": \\"string\\"
        }
    ]
}
\`\`\`

### **New User**

> **\`POST\`**  **/newUser**

**Request Body**

\`\`\`json
{
  \\"name\\": \\"string\\",
  \\"password\\": \\"string\\"
}
\`\`\`

**Response**

\`\`\`json
{
    \\"code\\": 200
}
\`\`\`
"
`;

exports[`Name of the group should  2`] = `
"# APIS

## About

## API

<!--api-start-->
### **Search**

Search for user

> **\`GET\`**  **/search?q={string}&page={number}&perPage={number}**

**Response**

\`\`\`json
{
    \\"code\\": 200,
    \\"posts\\": [
        {
            \\"name\\": \\"string\\"
        },
        {
            \\"name\\": \\"string\\"
        }
    ]
}
\`\`\`

### **New User**

> **\`POST\`**  **/newUser**

**Request Body**

\`\`\`json
{
  \\"name\\": \\"string\\",
  \\"password\\": \\"string\\"
}
\`\`\`

**Response**

\`\`\`json
{
    \\"code\\": 200
}
\`\`\`

<!--api-end-->
"
`;
